preprocess/docenc_main.cc

```diff
@@ -0,0 +1,258 @@
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <unistd.h>
+#include <vector>
+#include "util/file_stream.hh"
+#include "util/file_piece.hh"
+#include "util/string_piece.hh"
+#include "preprocess/base64.hh"
+
+using namespace std;
+
+enum Mode {
+	COMPRESS,
+	DECOMPRESS
+};
+
+string prefix_lines(string const &input, string const prefix)
+{
+	stringstream sin(input);
+	stringstream sout;
+
+	string line;
+	while (getline(sin, line))
+		sout << prefix << line << '\n';
+
+	return sout.str();
+}
+
+size_t decode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices, bool print_document_index, bool &delimiter_encountered) {
+	size_t document_index = 0;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+
+	for (StringPiece line : in) {
+		++document_index;
+
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+			}
+		}
+
+		string document;
+		preprocess::base64_decode(line, document);
+
+		if (!delimiter_encountered && document.find(delimiter == '\n' ? string("\n\n") : string(&delimiter, 1)) != string::npos)
+			delimiter_encountered = true;
+
+		if (print_document_index)
+			document = prefix_lines(document, to_string(document_index) + "\t");
```

Comment:
Why are we building a string `document` using `stringstream` only to output the string to a different stream?  Wouldn't it be easier and more efficient to just write to `out`?

Reply:
Honest answer? Because it was a bit of a lazy hack for a feature I needed at the time. docenc isn't used in any critical process. It can use some performance improvements.